### ISSUE TITLE
Editor crash when opening project stored in LUKS volume after machine reboot

### ISSUE THREAD

**Issue by blujai831** [godot-jolt/godot-jolt#635]
On Manjaro x86-64 with ldd 2.38, if you create a project in Godot 4.1.1, install Godot Jolt 0.8.0 from the in-editor asset library browser, close the project, reboot the computer, and try to open the project again, the editor will immediately crash.

Installing the debug symbols and running in gdb via `gdb --args $path_to_godot_binary --disable-crash-handler --path $path_to_project -e` will show the following as the cause:
```
Thread 1 "godot.linuxbsd." received signal SIGILL, Illegal instruction.
0x00007ffff25e1320 in _GLOBAL__sub_I_locale_inst.cc ()
   from $path_to_project/addons/godot-jolt/linux/godot-jolt_linux-x64_editor.so
```
Examining the backtrace will reveal the problem occurs when dynamically linking `godot-jolt_linux-x64_editor.so`.

The problem can be solved temporarily by deleting `$path_to_project/addons/godot-jolt` and reinstalling the plugin, but will recur after another machine reboot.

The minimal repro example does not contain the debug symbols, as they were too big. Installing them after extraction will be necessary to get a meaningful backtrace when reproducing the issue, assuming the issue is reproducible at all. The archive root contains a file `typescript` which is a shell session dump that includes a backtrace.

[godot-jolt-issue-repro.zip](https://github.com/godot-jolt/godot-jolt/files/12851193/godot-jolt-issue-repro.zip)

**Comment by mihe** [comment]
From what I understand you were using a source-built version of Godot and not the official builds. This leads me to a couple of questions:

1. Have you modified the engine in any way? If so, what are those changes?
2. Are you able to reproduce the crash with the official builds of Godot?
3. What flags are/were you using when building Godot from source?
4. Does it make any difference if you compile Godot with `use_static_cpp=yes` (or `production=yes` which also includes this)

**Comment by blujai831** [comment]
> From what I understand you were using a source-built version of Godot and not the official builds. This leads me to a couple of questions:
> 
>     1. Have you modified the engine in any way? If so, what are those changes?
It looks like pamac gave me an AUR package. Here's its PKGBUILD: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=godot-mono
On a cursory inspection I don't see anything in there that looks like a source patch.
>     2. Are you able to reproduce the crash with the official builds of Godot?
I've now tested with the following official builds (from the website's download archive, https://godotengine.org/download/archive):
- 4.1.1 stable
- 4.1.1 stable mono
- 4.1.2 stable
- 4.1.2 stable mono
The issue occurs with all of them.
>     3. What flags are/were you using when building Godot from source?
It looks like the PKGBUILD uses `platform=linuxbsd tools=yes module_mono_enabled=yes mono_glue=no`. From what I can tell, it then has the result build mono glue for itself.
>     4. Does it make any difference if you compile Godot with `use_static_cpp=yes` (or `production=yes` which also includes this)?
Just checked: it does not.

**Comment by mihe** [comment]
So I've tried running it on a Manjaro "Live USB" environment now, and it seems to work just fine for me. I tried the official builds, the [godot](https://archlinux.org/packages/extra/x86_64/godot/) package, as well as the AUR [godot-mono](https://aur.archlinux.org/packages/godot-mono) package.

I'm not quite sure how to proceed with this one. I have a feeling that the solution proposed in #378 would resolve this, but I can't be totally sure.

Any chance you're able to run things on another distro to see if things work there?

**Comment by blujai831** [comment]
I tried in an Ubuntu 22.04.3 LiveCD QEMU KVM instance. The command I used was `qemu-system-x86_64 -enable-kvm -smp 4 -m 8G -cdrom Downloads/ubuntu-22.04.3-desktop-amd64.iso`.

Bizarrely, the same issue occurs in the QEMU instance:

![image](https://github.com/godot-jolt/godot-jolt/assets/100004131/6b74d4b3-1848-4bd4-9fb6-f0ff9a93e013)

At this time I can't test in a native boot of the LiveCD, as I don't have a free pendrive on hand. Nor can I test in a non-KVM instance, as it then runs obscenely slowly. I'll get back to you when I'm able to test in a native boot.

**Comment by blujai831** [comment]
OK, with a native boot, the issue also occurs:

![godot-jolt-issue-in-ubuntu-livecd-native-boot](https://github.com/godot-jolt/godot-jolt/assets/100004131/f3110263-7168-454c-b5f8-252e7e0d5e09)

I'm beginning to wonder if perhaps the issue is my CPU. The crash is caused by SIGILL, after all.

`/proc/cpuinfo` reports my CPU as `AMD Ryzen 7 3750H with Radeon Vega Mobile Gfx`, 8-core.

It would be strange if that were the problem, though, since I can fix the issue until next reboot by deleting and redownloading the addon.

I've just noticed the .so in question differs between a fresh download and the .so in my project directory. Does Godot Jolt self-modify its shared library at some point?

The files are the same size. Attached is the result of `cmp -l` between them. `wc` reports 122398 lines of output from `cmp -l`, suggesting ~120KiB difference, or ~2%.

I wonder if maybe either Godot Jolt or Godot itself somehow auto-optimizes the .so, and does so in a manner that makes possibly incorrect assumptions about the host architecture rather than checking? Though it's bizarre this would result in a .so of the same size as the input. Also why would this only happen if the computer has been rebooted since last launch? Very weird.

[godot-jolt-so-cmp-l.txt](https://github.com/godot-jolt/godot-jolt/files/12908139/godot-jolt-so-cmp-l.txt)

**Comment by mihe** [comment]
> I'm beginning to wonder if perhaps the issue is my CPU. The crash is caused by SIGILL, after all.

I'd be curious to know what exact instruction it's freaking out over. Both Godot and Godot Jolt only target SSE2, which is ancient enough to be included in x86-64 natively, and none of the newer instruction sets, so I can't imagine the problem being that the binaries are compiled with instructions that something as new as a Ryzen 7 3750H can't handle.

> I've just noticed the .so in question differs between a fresh download and the .so in my project directory. Does Godot Jolt self-modify its shared library at some point?

Definitely not. That sounds bizarre. I'm 99% certain Godot itself doesn't mess with the library either. The only thing that springs to mind is maybe GDB messing with it in order to establish/fulfill the debug link between the library and the debug symbols.

Maybe we're missing the forest for all the trees. Given that it's crashing in `_GLOBAL__sub_I_locale_inst.cc`, what's your output when running `locale -a`?

**Comment by blujai831** [comment]
Good thinking. These are my locales according to `locale -a`:
```
C
C.utf8
en_US.utf8
POSIX
```

I've tried setting LANG to each of them individually. Same result for each of them

**Comment by mihe** [comment]
Apologies for the radio silence here. The locale stuff you posted looked normal, and I haven't really had any more ideas for how to proceed with this one.

I realize it's a couple of weeks now, but are you still up for trying to find out what's causing this?

If so, I'd be curious to uncover exactly what instruction is causing this `SIGILL, Illegal instruction` signal to fire, which I believe GDB should be able to tell us through the `disassemble /m` command once you've actually crashed within GDB. I'd also be curious what the `siginfo` command in GDB outputs at that time.

**Comment by blujai831** [comment]
Hey. No problem, I appreciate you still being willing to look into this even though it seems to only happen to me(!?). Here is a log of what happens when I do that. (`siginfo` on its own did not seem to be an available command, so I looked it up and tried `p $_siginfo` instead. I hope that was the right one.)
[typescript.txt](https://github.com/godot-jolt/godot-jolt/files/13481996/typescript.txt)

**Comment by blujai831** [comment]
It would have been using the already corrupted file, yes, though I didn't know at the time that that was what was going on.

At this point, I highly doubt this is actually a bug in Godot Jolt. The fault lies almost certainly either with dm-crypt, cryptsetup, or myself in the way I'm using them. (I doubt it's a hardware issue, since the problem presents in the same way every time.) I'm way out of my depth in that respect, but I _naively_ suspect _maybe_ it has something to do with storing a shared library in an encrypted volume, and then loading it, and then shutting down, and, I don't know, not being able to finish encrypting the shared library because it's still paged or something, I really have no clue.

I'm sorry to send you on this wild goose chase. Do you think it's a safe bet to close the issue at this point?

**Comment by mihe** [comment]
> I'm way out of my depth in that respect, but I _naively_ suspect _maybe_ it has something to do with storing a shared library in an encrypted volume, and then loading it, and then shutting down, and, I don't know, not being able to finish encrypting the shared library because it's still paged for some reason or something, I really have no clue.

I can't say I have a clue either. I have little to no experience with disk encryption, especially on Linux.

> Do you think it's a safe bet to close the issue at this point?

Does this happen with other native extension as well? [Debug Draw 3D](https://godotengine.org/asset-library/asset/1766) is one that I use myself from time to time. I'd be curious to know if you experience the same issue when using it.

**Comment by blujai831** [comment]
It didn't seem to happen with Debug Draw 3D.

I tried to capture what was going on with strace, but just as I started trying that, I stopped being able to reproduce the issue at all, so of course the resulting logs didn't tell me anything useful (can't trace something that isn't happening). I regret that before the issue stopped occurring, I couldn't at the very least isolate the moment the library is corrupted (re. whether it happens before closing, after closing but before reboot, after reboot but before reopening and crashing, or after reopening and crashing). I would have liked to do that, and will if I can get the issue to happen again. It was happening consistently until now.

**Comment by mihe** [comment]
> To clarify, when I had Debug Draw 3D installed in parallel with Godot Jolt, it happened with Godot Jolt, but not with Debug Draw 3D.

By "it happened" you mean that the Godot Jolt binary was corrupted, but Debug Draw 3D was left untouched?

**Comment by blujai831** [comment]
Yes, that's right.

**Comment by mihe** [comment]
So strange.

I really don't see what I could be doing that would be so crazy as to freak out disk encryption of all things.

**Comment by mihe** [comment]
So I installed Ubuntu with its full disk encryption enabled, which according to `cryptsetup isLuks` seems to be using LUKS as well, and I was not able to reproduce this issue in that particular setup.

I also decided to take a quick look in Ghidra to see exactly what's being overwritten at the offsets you provided from `cmp -l`, and I can't really see any immediate rhyme or reason to the chunks being overwritten.

I frankly don't know how to proceed with this one, and seeing as how nobody else has piled on here in the months that this issue has been open I've decided to close this one as an unsolved mystery for now.
